tests/utils/test_process_util.py

```diff
@@ -32,6 +36,101 @@ def tearDown(self):
         if self.tmp_dir is not None:
             shutil.rmtree(self.tmp_dir)
 
+    def test_wait_for_process_completion_or_timeout_should_terminate_cleanly(self):
+        process = subprocess.Popen(
+            "date",
+            shell=True,
+            cwd=self.tmp_dir,
+            env={},
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE)
+
+        timed_out, ret = wait_for_process_completion_or_timeout(process, 5)
+        self.assertEquals(timed_out, False)
+        self.assertEquals(ret, 0)
+
+    def test_wait_for_process_completion_or_timeout_should_kill_process_on_timeout(self):
+        process = subprocess.Popen(
+            "sleep 4s",
+            shell=True,
+            cwd=self.tmp_dir,
+            env={},
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            preexec_fn=os.setsid)
+
+        with patch('azurelinuxagent.common.utils.processutil.os.killpg') as mock_kill:
+            timed_out, ret = wait_for_process_completion_or_timeout(process, 2)
```

Comment:
2 seconds is too long for a unit test.

if you add some synchronization with the child process you may avoid the sleep altogether. For example, the child process creates a file and then sleeps for a minute or so.

the test waits for the file to be created and then it just uses an arbitrary timeout. You don't mock kill, but mock sleep. Then you assert that sleep was called the expected number of times (i.e. we waited for the timeout period before killing the process) and that the process was terminated (i.e. the process was killed after the timeout elapsed).

Reply:
As discussed, this test doesn't even need to synchronize with the child process since all it needs is the process to have already started. I'm mocking sleep and asserting we're "waiting" before killing the process. I've removed the mock for the kill to make the test more end-to-end.